core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java

```diff
@@ -71,19 +70,74 @@ public WrapperQueryBuilder(byte[] source, int offset, int length) {
      */
     public WrapperQueryBuilder(BytesReference source) {
         this.source = source.array();
-        this.offset = source.arrayOffset();
-        this.length = source.length();
+    }
+
+    public byte[] source() {
+        return this.source;
+    }
+
+    @Override
+    public WrapperQueryBuilder boost(float boost) {
+        //no-op: WrapperQueryBuilder doesn't support boost, we should be consistent and ignore it here too.
+        return this;
+    }
+
+    @Override
+    public WrapperQueryBuilder queryName(String queryName) {
+        //no-op: WrapperQueryBuilder doesn't support _name, we should be consistent and ignore it here too.
+        return this;
+    }
+
+
+    @Override
+    public String getName() {
+        return NAME;
     }
 
     @Override
     protected void doXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(NAME);
-        builder.field("query", source, offset, length);
+        builder.field("query", source);
         builder.endObject();
     }
 
     @Override
-    public String getName() {
-        return NAME;
+    protected Query doToQuery(QueryParseContext parseContext) throws IOException {
+        try (XContentParser qSourceParser = XContentFactory.xContent(source).createParser(source)) {
+            final QueryParseContext context = new QueryParseContext(parseContext.index(), parseContext.indexQueryParserService());
+            context.reset(qSourceParser);
+            Query query = context.parseInnerQuery();
+            parseContext.combineNamedQueries(context);
```

Comment:
parsing in toQuery feels very weird here... it will be a problem especially when we split the contexts into "parsing" and "to_query" because we will need the former here while we will have only the latter. @cbuescher what do you think we should do here instead? I am not sure yet....